### PR TITLE
doc: Add design document

### PR DESCRIPTION
**Pull request by asymmetric**
Explains:

* Basic operations
* Custom versioning
* Sub-resources
* `peer-watch-file` `ConfigMap`

Signed-off-by: Lorenzo Manacorda <lorenzo@kinvolk.io>

Closes #253.

### CODE REVIEW

File: doc/design.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+arguments, the operator resorts to using a `ConfigMap` instead.  `ConfigMap`s
+can be mounted as files inside `Pod`s, and Habitat services can be started with
+the `--peer-watch-file` flag, which expects a file.  This allows us to start
+*all* `Pod`s in the `StatefulSet` with the same set of arguments.
+
+The `ConfigMap` is maintained by the operator in the following ways:
+
+* When it's empty, it populates it with the IP of one of the running `Pod`s (the
+  so called "leader")
+* Whenever a leader dies, the operator replaces the IP in the `ConfigMap` with
```
Comment (krnowak) [inline doc/design.md]: Whenever the `Pod` dies

File: doc/design.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  that of another running Pod, which becomes the new leader
```
Comment (krnowak) [inline doc/design.md]: Pod in backticks?

File: doc/design.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+The Habitat operator manages Habitat services on Kubernetes using the operator
+pattern. During the initial run it registers the [Custom Resource Definition
+(CRD)][crd] for Habitat. The Habitat `CRD` is essentially the schema that
+describes the contents of the manifests for deploying individual Habitat objects
+using `StatefulSet`s.
+
+Once the operator is running, it performs the following actions:
+
+* watches for new `Habitat` manifests and deploys corresponding sub-resources
```
Comment (krnowak) [inline doc/design.md]: In bullet points I prefer to start with a capital letter and end with a full dot.

You may have a different preference, but please make it at least consistent.

File: doc/design.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+The `ConfigMap` is maintained by the operator in the following ways:
+
+* When it's empty, it populates it with the IP of one of the running `Pod`s (the
+  so called "leader")
+* Whenever a leader dies, the operator replaces the IP in the `ConfigMap` with
+  that of another running Pod, which becomes the new leader
+
+## Sub-resources
+
+As mentioned above, the operator creates several resources whenever it detects a
```
Comment (krnowak) [inline doc/design.md]: That makes me think that maybe we should put a label on all the resources we create, so we could tell the users in this document how to recognize a resource that is managed by the operator.

I think we don't put any label on on Deployments and StatefulSets. We do it for ConfigMaps and Pods.

Comment (asymmetric) [reply doc/design.md]: We do for [Deployments](
https://github.com/kinvolk/habitat-operator/tree/914de680153a3b83fe3ee4a9e7303cd8248bbf2a/pkg/controller/v1beta1/controller.go#L621) and [StatefulSets](
https://github.com/kinvolk/habitat-operator/tree/914de680153a3b83fe3ee4a9e7303cd8248bbf2a/pkg/controller/v1beta2/stateful_sets.go#L97) as well.

I'll add a mention of it here.

File: doc/design.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+* periodically checks running Habitat resources against the manifests and acts
+  on the differences found
+
+For instance, when the user creates a new custom object of type Habitat by
+submitting a new manifest with `kubectl`, the operator fetches that object and
+creates the corresponding Kubernetes structures (`StatefulSet`s, `Deployment`s,
+`ConfigMap`s, `Secret`s, `PersistentVolume`s) according to its definition.
+
+Another example is changing the Docker image inside a Habitat object. In this
+case, the operator first goes to all `StatefulSet`s it manages and updates them
```
Comment (indradhanush) [inline doc/design.md]: s/`Statefultset`s/`Statefulsets` makes it easier to read in the rendered version.

File: doc/design.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+with the new Docker images; afterwards, all `Pod`s from each `StatefulSet` are
```
Comment (indradhanush) [inline doc/design.md]: Same for "`Pod`s" and other similar places.

File: doc/design.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+killed one by one ([rolling update][rolling]) and the replacements are spawned
+automatically by each `StatefulSet` with the new Docker image.
+
+## Service groups
+
+In a non-Kubernetes Habitat setup, every service in a group after the first one
+is [started][hab-sg] with the `--peer` flag, pointing to the first service's IP.
+
+Since it's not possible in Kubernetes to launch some members with different
```
Comment (indradhanush) [inline doc/design.md]: members of?

File: doc/design.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+arguments, the operator resorts to using a `ConfigMap` instead.  `ConfigMap`s
+can be mounted as files inside `Pod`s, and Habitat services can be started with
+the `--peer-watch-file` flag, which expects a file.  This allows us to start
+*all* `Pod`s in the `StatefulSet` with the same set of arguments.
+
+The `ConfigMap` is maintained by the operator in the following ways:
+
+* When it's empty, it populates it with the IP of one of the running `Pod`s
+* Wenever a `Pod` dies, it makes sure that the IP in the `ConfigMap` is still
```
Comment (indradhanush) [inline doc/design.md]: Typo in "Wenever".

File: doc/design.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+Since it's not possible in Kubernetes to launch some members with different
+arguments, the operator resorts to using a `ConfigMap` instead.  `ConfigMap`s
+can be mounted as files inside `Pod`s, and Habitat services can be started with
+the `--peer-watch-file` flag, which expects a file.  This allows us to start
+*all* `Pod`s in the `StatefulSet` with the same set of arguments.
+
+The `ConfigMap` is maintained by the operator in the following ways:
+
+* When it's empty, it populates it with the IP of one of the running `Pod`s
```
Comment (indradhanush) [inline doc/design.md]: Maybe add a line or two on how the operator decides which `Pod` should become leader?

File: doc/design.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+assigned to a running `Pod`, and if not, replaces it
+
+## Sub-resources
+
+As mentioned above, the operator creates several resources whenever it detects a
+new `Habitat`. These sub-resources, all labeled with `habitat=true`, are to be
+considered read-only; any modifications to them will be overwritten by the
+operator.
+
+The proper way to make changes is **always** to make them on the Habitat object,
```
Comment (indradhanush) [inline doc/design.md]: Nit: "is to **always** make them"

File: doc/design.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+and let the operator figure out the actual steps it needs to take.
```
Comment (indradhanush) [inline doc/design.md]: This is not very clear to me as someone unfamiliar with the operator. When you say make changes on the `Habitat` object what are other ways by which things can be changed? Do you mean changing the `Pod` resource for example?

File: doc/design.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+## CRD Versioning
+
+According to [Semantic Versioning][semver], backwards incompatible changes require
+releasing a new major version. Unfortunately, it is [currently not
+possible][crd-vers] to run multiple versions of a CRD side-by-side. The
+work-around currently relies on two steps:
```
Comment (indradhanush) [inline doc/design.md]: "work-around" should be one word.

File: doc/design.md (hunk 9 above)
Comment (asymmetric) [reply doc/design.md]: There is no leader no more! :) (@krnowak was right that it would confuse people)

File: doc/design.md (hunk 11 above)
Comment (asymmetric) [reply doc/design.md]: Yes, that's the point of this section. Is that not clear enough? Suggestions?

File: doc/design.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+killed one by one ([rolling update][rolling]) and the replacements are spawned
+automatically by each `StatefulSet` with the new Docker image.
+
+## Service groups
+
+In a non-Kubernetes Habitat setup, every service in a group after the first one
+is [started][hab-sg] with the `--peer` flag, pointing to the first service's IP.
+
+Since it's not possible in Kubernetes to launch `Pod`s in a
+`Deployment`/`StatefulSet`with different arguments, the operator resorts to
```
Comment (krnowak) [inline doc/design.md]: Space between "StatefulSet" and "with".

File: doc/design.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+using a `ConfigMap` instead.  `ConfigMap`s can be mounted as files inside
+`Pod`s, and Habitat services can be started with the `--peer-watch-file` flag,
+which expects a file.  This allows us to start *all* `Pod`s in the `StatefulSet`
+with the same set of arguments.
+
+The `ConfigMap` is maintained by the operator in the following ways:
+
+* When it's empty, it populates it with the IP of one of the running `Pod`s
+(which `Pod` gets chosen is irrelevant, the important thing is that
+the`Pod` is in the `Running` state)
```
Comment (krnowak) [inline doc/design.md]: Space between "the" and "Pod".

File: doc/design.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+and let the operator figure out the actual steps it needs to take.
+
+## CRD Versioning
+
+According to [Semantic Versioning][semver], backwards incompatible changes require
+releasing a new major version. Unfortunately, it is [currently not
+possible][crd-vers] to run multiple versions of a CRD side-by-side. The
+workaround currently relies on two steps:
+
+* defining an additional field, `customVersion`
```
Comment (krnowak) [inline doc/design.md]: Capitalize "defining".

**Comment by asymmetric** [comment]
PTALAA.

File: doc/design.md (hunk 14 above)
Comment (krnowak) [reply doc/design.md]: This one was not fixed.

**Comment by krnowak** [comment]
LFAD, but squash some commits first, please.

**Pull request merged**